jms/src/main/scala/akka/stream/alpakka/jms/JmsSourceStage.scala

```diff
@@ -5,80 +5,319 @@
+      }
 
-      private val handleMessage = getAsyncCallback[Message](msg => {
-        require(queue.size <= bufferSize)
-        if (isAvailable(out)) {
-          pushMessage(msg)
-        } else {
-          queue.enqueue(msg)
+  private def abortSessions(ex: Throwable): Unit =
+    if (stopping.compareAndSet(false, true)) {
+      Future {
+        try {
+          jmsConnection.close()
+          log.info("JMS connection {} closed", jmsConnection)
+          markAborted.invoke(ex)
+        } catch {
+          case NonFatal(e) => log.error(e, "Error closing JMS connection {}", jmsConnection)
         }
-      })
+      }
+    }
 
-      override def preStart(): Unit =
-        initSessionAsync(getDispatcher)
+  private[jms] def killSwitch = new KillSwitch {
+    override def shutdown(): Unit = stopSessions()
+    override def abort(ex: Throwable): Unit = abortSessions(ex)
+  }
 
-      private def pushMessage(msg: Message): Unit = {
-        push(out, msg)
-        backpressure.release()
+  override def postStop(): Unit = {
+    queue.clear()
+    stopSessions()
+  }
+}
+
+class NoAckLogic(shape: SourceShape[Message],
+                 out: Outlet[Message],
+                 settings: JmsSourceSettings,
+                 attributes: Attributes)
+    extends SourceStageLogic[Message](shape, out, settings, attributes) {
+
+  private val bufferSize = (settings.bufferSize + 1) * settings.sessionCount
+
+  private val backpressure = new Semaphore(bufferSize)
+
+  private[jms] def createSession(connection: Connection, createDestination: Session => javax.jms.Destination) = {
+    val session =
+      connection.createSession(false, settings.acknowledgeMode.getOrElse(AcknowledgeMode.AutoAcknowledge).mode)
+    new JmsSession(connection, session, createDestination(session))
+  }
+
+  private[jms] def pushMessage(msg: Message): Unit = {
+    push(out, msg)
+    backpressure.release()
+  }
+
+  override private[jms] def onSessionOpened(jmsSession: JmsSession): Unit =
+    jmsSession
+      .createConsumer(settings.selector)
+      .onComplete {
+        case Success(consumer) =>
+          consumer.setMessageListener(new MessageListener {
+            def onMessage(message: Message): Unit = {
+              backpressure.acquire()
+              handleMessage.invoke(message)
+            }
+          })
+        case Failure(e) =>
+          fail.invoke(e)
       }
+}
+
+class ExplicitAckLogic(shape: SourceShape[AckEnvelope],
+                       out: Outlet[AckEnvelope],
+                       settings: JmsSourceSettings,
+                       attributes: Attributes)
+    extends SourceStageLogic[AckEnvelope](shape, out, settings, attributes) {
 
-      override private[jms] def onSessionOpened(): Unit =
-        jmsSession.createConsumer(settings.selector).onComplete {
+  private val maxPendingAck = settings.bufferSize
+
+  private[jms] def createSession(connection: Connection, createDestination: Session => javax.jms.Destination) = {
```

Comment:
I think a bunch of these can either be entirely `private` or `protected` (may be a left over of previous code rather than introduced in this PR but please look those through and adjust as needed)

Reply:
Will do. Thanks!

Reply:
Checked them. All implement or override the trait `JmsConnector` which I defined as `private[jms]`. For this reason they cannot be private. The call now is to make them `protected` or `private[jms]` which has the same effect but is just package private. Can you please confirm you really want to open this up for everyone to inherit/override? With confirmation I can change to `protected`. Thanks!